hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/processors/StreamToStreamJoinP.java

```diff
@@ -0,0 +1,233 @@
+                pendingOutput = joinedRow;
+            }
+            return false;
+        }
+
+        if (!iterator.hasNext()) {
+            iterator = null;
+            currItem = null;
+            return true;
+        }
+
+        JetSqlRow preparedOutput = ExpressionUtil.join(
+                ordinal == 0 ? currItem : iterator.next(),
+                ordinal == 0 ? iterator.next() : currItem,
+                joinInfo.condition(),
+                evalContext
+        );
+
+        if (preparedOutput != null && !tryEmit(preparedOutput)) {
+            assert pendingOutput == null;
+            pendingOutput = preparedOutput;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean tryProcessWatermark(int ordinal, @Nonnull Watermark watermark) {
+        // if pending watermarks available - try to send them
+        if (pendingWatermark != null && !tryEmit(pendingWatermark)) {
+            return false;
+        }
+        pendingWatermark = null;
+
+        // update wm state
+        offer(watermark);
+
+        // try to clear buffers if possible
+        clearExpiredItemsInBuffer(ordinal, watermark);
+
+        // We can't immediately emit current WM, as it could render items in buffers late.
+        // Instead, we can emit WM with the minimum available time for this WM key.
+        long minimumItemTime = findMinimumBufferTime(ordinal, watermark);
+        if (minimumItemTime != Long.MAX_VALUE) {
+            Watermark wm = new Watermark(minimumItemTime, watermark.key());
+            if (!tryEmit(wm)) {
+                assert pendingWatermark == null;
+                pendingWatermark = wm;
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private void offer(Watermark watermark) {
+        byte key = watermark.key();
+        final Map<Byte, Long> wmKeyMapping = postponeTimeMap.get(key);
+        for (byte i : wmKeyMapping.keySet()) {
+            Long curr = wmState.get(i).get(key);
+            wmState.get(i).put(key, (curr != null && curr != Long.MIN_VALUE)
+                    ? Math.min(watermark.timestamp() - wmKeyMapping.get(i), curr)
+                    : watermark.timestamp() - wmKeyMapping.get(i));
+        }
+    }
+
+    private long findMinimumBufferTime(int ordinal, Watermark watermark) {
+        byte key = watermark.key();
+        ToLongFunctionEx<JetSqlRow> extractor = ordinal == 0 ? leftTimeExtractors.get(key) : rightTimeExtractors.get(key);
+
+        long min = Long.MAX_VALUE;
+        for (JetSqlRow row : buffer[ordinal]) {
+            min = Math.min(min, extractor.applyAsLong(row));
+        }
+        return min;
+    }
+
+    private long findMinimumGroupTime(byte group) {
+        long min = Long.MAX_VALUE;
+        for (long i : wmState.get(group).values()) {
+            min = Math.min(min, i);
+        }
+        return min;
+    }
+
+    private void clearExpiredItemsInBuffer(int ordinal, Watermark wm) {
```

Comment:
This is not correct. We can remove an item from the buffer after it is late according to ALL postponed watermarked fields in it. If there are two WMs and it's late according to one of them, we still can receive a row from the other input that could join with it.

This might do it, but some tests fail, I hope it's correct:
```java
    private void clearExpiredItemsInBuffer(int ordinal) {
        // remove items that are late according to all watermarks in them. Run after the wmState was changed.
        Map<Byte, ToLongFunctionEx<JetSqlRow>> extractors = ordinal == 0 ? leftTimeExtractors : rightTimeExtractors;
        ToLongFunctionEx<JetSqlRow>[] extractors2 = new ToLongFunctionEx[extractors.size()];
        long[] limits = new long[extractors.size()];
        int i = 0;
        for (Entry<Byte, ToLongFunctionEx<JetSqlRow>> entry : extractors.entrySet()) {
            extractors2[i] = entry.getValue();
            limits[i] = findMinimumGroupTime(entry.getKey());
            assert limits[i] != Long.MAX_VALUE;
            i++;
        }
        buffer[ordinal].removeIf(row -> {
            for (int j = 0; j < extractors2.length; j++) {
                if (extractors2[j].applyAsLong(row) >= limits[j]) {
                    return false;
                }
            }
            return true;
        });
    }
```
Naming in my code isn't great :)